lib/java/persistance/src/main/java/org/enso/persist/PerGenerator.java

```diff
@@ -36,6 +39,23 @@ static byte[] writeObject(Object obj, Function<Object, Object> writeReplace) thr
     arr[9] = (byte) ((at >> 16) & 0xff);
     arr[10] = (byte) ((at >> 8) & 0xff);
     arr[11] = (byte) (at & 0xff);
+
+    var counts = g.knownTypes;
+    var list = new ArrayList<>(counts.entrySet());
+    list.sort(
+        (a, b) -> {
+          return a.getValue()[0] - b.getValue()[0];
+        });
+
+    for (var i = 0; i < list.size(); i++) {
+      if (i == 30) {
+        break;
+      }
```

Comment:
It seems like the same logic is repeated here. Another reason to abstract it somehow

Reply:
When I started my exploration of the cache files, I first tried to see what's being _deserialized_. However that wasn't very useful:
- we deserialize _on demand_ - e.g. lazily - can be workarounded by `ir.preorder()`, but that's not exactly what's happening
- it is hard to find out whether an object was read again or is an existing reference
- impossible to find out how many bytes are occupied by an object

As such I decided to add the logging into `PerGenerator` and check what is being _serialized_. The deserialization logic can now be removed - done in 65ff16c3d2

Reply:
Isolated into `Histogram` class in 015f4bb09b462c275ce895aa418370500fbc3b84